lib/reader/src/parser.rs

```diff
@@ -1086,8 +1104,29 @@ impl<'a> Parser<'a> {
         // Pass a reference to `ctx.values` instead of `ctx` itself since the `Values` iterator
         // holds a reference to `ctx.function`.
         self.add_values(&mut ctx.map,
-                        results.into_iter(),
-                        ctx.function.dfg.inst_results(inst))
+                        results.iter().map(|v| *v),
+                        ctx.function.dfg.inst_results(inst))?;
+
+        if let Some(ref result_locations) = result_locations {
+            if result_locations.len() != results.len() {
+                return err!(self.loc,
+                            "must have same number of result locations as results");
+            } else {
+                for (&src_value, &loc) in results.iter().zip(result_locations) {
```

Comment:
Since the real result values are on the instruction now, you can use:

```rust
for (&value, &loc) in ctx.function.dfg.inst_results(inst).zip(result_locations) {
```

Then you don't need the `ctx.map.get_value` lookup below.